### DISCUSSION TITLE
cli: how to globally disable scientific notation output?

### DISCUSSION THREAD

**Discussion by aborruso** [duckdb/duckdb#12274] [Q&A]
Hi,
I'm using postgres extension to connect to a table in which I have a double field.

If I run a select to output a CSV

```
duckdb --csv -c "ATTACH 'dbname=dati_m user=postgres host=37.37.37.37 password=asdasdsdas' AS db (TYPE POSTGRES, READ_ONLY);select my_field from db.schema.my_dataset"
```

I have this output

```
my_field 
7.94e-05
0.000104
5.31e-05
9.18e-05
0.000274
2.73e-07
```

Is there a global way to disable scientific notation output without hardcoding the output for individual field names, so that floating point numbers are displayed as shown in the image below?

Thank you

![image](https://github.com/duckdb/duckdb/assets/30607/521d6e59-2878-424b-bc3c-854749776f2e)

**Comment by Tishj** [comment]
I'm not sure there is, a possible work around would be to use format, I presume that can disable scientific notation

**Reply by Tishj** [reply]
```
D select 0.00000794::DOUBLE as data, format('{:.54f}', data);
┌──────────┬──────────────────────────────────────────────────────────┐
│   data   │                format('{:.54f}', "data")                 │
│  double  │                         varchar                          │
├──────────┼──────────────────────────────────────────────────────────┤
│ 7.94e-06 │ 0.000007940000000000000212447247294189622834892361424863 │
└──────────┴──────────────────────────────────────────────────────────┘
```

**Reply by aborruso** [reply]
@Tishj I know this and thank you.

But mine was just an example. I have so many columns and then in general to me it seems strange and it seems like a limitation that there is no way to go out globally as a simple float :(

**Reply by Tishj** [reply]
I think it's probably a side effect of using `fmt` format internally
```c++
template <>
string_t StringCast::Operation(double input, Vector &vector) {
	std::string s = duckdb_fmt::format("{}", input);
	return StringVector::AddString(vector, s);
}
```

**Reply by aborruso** [reply]
Hi @Tishj thank you again I have no idea on how to use this code.

I'm not a developer, but simply a stupid user :)

I think I will open an issue on this, because I think it is not good to have this

```
duckdb --csv -c "select 0.00000794::DOUBLE as data;"

data
7.94e-06
```

I think by default you should have

```
data
0.00000794
```

What do you think?

Thanks again

**Reply by aborruso** [reply]
I have created a feature request https://github.com/duckdb/duckdb/discussions/12279

**Reply by Tishj** [reply]
It means that the library we bundle to format output `fmt` by default uses scientific notation for doubles.
It does not seem to be possible to disable this without specifying a precision to use, such as the example I showed you

The best we can likely do without disturbing the existing code too much is add an option to provide a default format to use for doubles - which would require a precision

Essentially turning `select dbl from tbl` into `select format(<default_double_format>, dbl) from tbl` implicitly

**Reply by FrancoisLepoutre** [reply]
Hi Thijs, thanks for looking into the matter. :-)

Among duckdb users there are developers with IT-background and an application or at least automation of some sort of a repeatable data exploration scheme in mind. Those will indeed value accuracy above precision. 

IMHO **Andrea is have a plain great point here in terms of usability. End-users who care about the data itself will indeed value precision over accuracy!**

**Reply by aborruso** [reply]
@FrancoisLepoutre thank you very much and thank you again @Tishj 

In some way `0.00000794` and `7.94e-06` have the same accuracy and precision, but they have a different representation (notation).

In my use case I have this kind of values, in different columns 0.0000531, 0.000000273, 0.00863, ecc.. Here we have the same significant figures but a different order of magnitude.

Suppose I started from a CSV file and imported everything into duckdb, and in my CSV the data was in decimal notation. Imagine that I had 10 columns. Now if, after my processing in duckdb, I want to export back to decimal notation, I will have to set the formatting for each column.
Moreover if my start CSV is a bad CSV in which for the same column I have values with a different number of significant figures, if I set `{:.8f}` I will have `0.00000794` for `7.94e-06` and `0.00079400` for `7.94e-04`.

And I think that would be a good feature to enable intelligent conversion between scientific and decimal notation in the same column, ensuring consistency in significant figures, without unnecessary trailing zeros.

I hope I have explained myself and once again, thank you very much.

**Reply by Tishj** [reply]
These are the available formatting options:
<https://fmt.dev/6.1.1/syntax.html#format-specification-mini-language>

If you can find one that does what you want we can add an option to change the default to that (if the option is set), but I think that's the only chance of making this change

Perhaps the `f` format does what you want?
<https://github.com/fmtlib/fmt/issues/2531#issuecomment-933794168> seems to hint at that

**Reply by aborruso** [reply]
@Tishj once again thank you.

And `f` is the way!!

It would be a great feature to have a global way, to have an option, to set it for the output of all double fields of a table.

**Reply by aborruso** [reply]
Hi @Tishj is it useful to do something? Should I open a feature request?

Thank you

**Comment by simran99** [comment]
Hi

Had few observations, please add/correct if I am missing something

It seems like `{:f}` can not be relied upon as it converts into float then varchar, hence losing precision 

Say if we would have a global flag to switch off scientific notation, it would be use for such cases 

```
 select format('{:8f}',    CAST('0.000000000000007' AS double) ); -- CAST('0.000000000000007' AS double) results into 7e-15
┌──────────────────────────────────────────────────────┐
│ format('{:8f}', CAST('0.000000000000007' AS DOUBLE)) │
│                       varchar                        │
├──────────────────────────────────────────────────────┤
│ 0.000000                                             │
└──────────────────────────────────────────────────────┘
```


```
select format('{:f}',    CAST('0.9999999999997' AS double) );   -- loss of precision 
┌───────────────────────────────────────────────────┐
│ format('{:f}', CAST('0.9999999999997' AS DOUBLE)) │
│                      varchar                      │
├───────────────────────────────────────────────────┤
│ 1.000000                                          │
└───────────────────────────────────────────────────┘

select CAST('0.9999999999997' AS double)::FLOAT;  
┌──────────────────────────────────────────────────┐
│ CAST(CAST('0.9999999999997' AS DOUBLE) AS FLOAT) │
│                      float                       │
├──────────────────────────────────────────────────┤
│                                              1.0 │
└──────────────────────────────────────────────────┘

select CAST('0.9999999999997' AS double);  --expected 
┌───────────────────────────────────┐
│ CAST('0.9999999999997' AS DOUBLE) │
│              double               │
├───────────────────────────────────┤
│                   0.9999999999997 │
└───────────────────────────────────┘
```

Also Varchar that is a double depicted in scientific notation when casted to decimal results into incorrect result.

```

select CAST('7e-15' as DECIMAL(38,10)) ;  ---INCORRECT, rounding off even though not expected 
┌─────────────────────────────────┐
│ CAST('7e-15' AS DECIMAL(38,10)) │
│         decimal(38,10)          │
├─────────────────────────────────┤
│                    0.0000000001 │
└─────────────────────────────────┘

 select CAST('0.000000000000007' as DECIMAL(38,10)) ;  --CORRECT, consistent with trino/athena
┌─────────────────────────────────────────────┐
│ CAST('0.000000000000007' AS DECIMAL(38,10)) │
│               decimal(38,10)                │
├─────────────────────────────────────────────┤
│                                0.0000000000 │
└─────────────────────────────────────────────┘

```
 Have raised the same here - https://github.com/duckdb/duckdb/issues/14779#issuecomment-2477923836
